### PR TITLE
Add support for 'hs2019' algorithm

### PR DESCRIPTION
**Pull request by sebastien-rosset**
In the most recent [IETF draft](https://datatracker.ietf.org/doc/draft-ietf-httpbis-message-signatures/), the `algorithm` field should be set to HS2019.
This PR adds support for HS2019

There is a dependency on #31 on #32.

### CODE REVIEW

File: src/main/java/org/tomitribe/auth/signatures/Signature.java [hunk 1]
```diff
@@ -83,32 +100,48 @@
...
+     * @param signingAlgorithm An identifier for the HTTP Signature algorithm.
+     *  This should be "hs2019" except for legacy applications that use an older version of the draft HTTP signature specification.
+     * @param algorithm The detailed algorithm used to sign the message.
      * @param parameterSpec optional cryptographic parameters for the signature.
-     * @param algorithm The algorithm used to sign the message.
      * @param headers The list of HTTP headers that will be used in the signature.
      */
-    public Signature(final String keyId, final String algorithm, final AlgorithmParameterSpec parameterSpec, final String... headers) {
-        this(keyId, getAlgorithm(algorithm), parameterSpec, null, Arrays.asList(headers));
+    public Signature(final String keyId, final String signingAlgorithm, final String algorithm, final AlgorithmParameterSpec parameterSpec, final List<String> headers) {
```
Comment (jgallimore) [inline src/main/java/org/tomitribe/auth/signatures/Signature.java:109]: I wonder if we can keep the old constructors and mark them with `@Deprecated` in order to maintain some backwards compatibility for other clients?

File: src/main/java/org/tomitribe/auth/signatures/SigningAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+     * 
+     * Recommend support for:
+     *   RSASSA-PSS [RFC8017] using SHA-512 [RFC6234]
+     *   HMAC [RFC2104] using SHA-512 [RFC6234]
+     *   ECDSA using curve P-256 [DSS] and SHA-512 [RFC6234]
+     *   Ed25519ph, Ed25519ctx, and Ed25519 [RFC8032]
+     */
+    HS2019("hs2019", null),
+
+    // Deprecated, SHA-1 is not secure.
```
Comment (jgallimore) [inline src/main/java/org/tomitribe/auth/signatures/SigningAlgorithm.java:55]: Do you think we should mark these with `@Deprecated`?

File: src/main/java/org/tomitribe/auth/signatures/SigningAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ * If algorithm is provided and differs from or is incompatible with the algorithm
+ * or key material identified by keyId (for example, algorithm has a value of
+ * rsa-sha256 but keyId identifies an EdDSA key), then a verification exception is
+ * raised.
+ * 
+ * The default value for this parameter should be "hs2019".
+ * 
+ * @see https://www.ietf.org/id/draft-ietf-httpbis-message-signatures-00.html
+ */
+public enum SigningAlgorithm {
```
Comment (jgallimore) [inline src/main/java/org/tomitribe/auth/signatures/SigningAlgorithm.java:41]: I keep wondering whether there's a way to have one enum as opposed to two. I'm not suggesting that we change this, more just thinking out loud.

File: src/main/java/org/tomitribe/auth/signatures/Signature.java (hunk 1 above)
Comment (jgallimore) [reply src/main/java/org/tomitribe/auth/signatures/Signature.java:109]: I suspect String... headers makes keeping the old constructors not possible.

File: src/main/java/org/tomitribe/auth/signatures/SigningAlgorithm.java (hunk 2 above)
Comment (sebastien-rosset) [reply src/main/java/org/tomitribe/auth/signatures/SigningAlgorithm.java:55]: Sure, that makes sense.

File: src/main/java/org/tomitribe/auth/signatures/SigningAlgorithm.java (hunk 3 above)
Comment (sebastien-rosset) [reply src/main/java/org/tomitribe/auth/signatures/SigningAlgorithm.java:41]: I've been wondering the same thing for python and golang, but I think in the draft they have decoupled the signature algorithm identifier from the details of the cryptographic algorithms, so it seems inevitable.

**Comment by sebastien-rosset** [comment]
Thank you for merging.

**Comment by jgallimore** [comment]
@sebastien-rosset just wanted to let you know that org.tomitribe:tomitribe-http-signatures:1.4 has been released, and should be making its way to Maven Central. Thanks again for your PRs.

**Pull request merged**